File: 2024/128. Longest Consecutive Sequence.cpp
class Solution {
public:
    int find(unordered_map<int, int>& parents, int x) {
        if (parents[x] != x) {
            parents[x] = find(parents, parents[x]);
        }
        return parents[x];
    }

    void merge(unordered_map<int, int>& parents, unordered_map<int, int>& ranks, int x, int y) {
        int px = find(parents, x), py = find(parents, y);
        if (px == py) {
            return;
        }
        if (ranks[px] > ranks[py]) {
            parents[py] = px;
            ranks[px] += ranks[py];
        } else {
            parents[px] = py;
            ranks[py] += ranks[px];
        }
    }
    int longestConsecutive(vector<int>& nums) {
        unordered_map<int, int> parents, ranks;
        for (int n : nums) {
            parents[n] = n;
            ranks[n] = 1;
        }
        for (int n : nums) {
            if (ranks.find(n - 1) != ranks.end()) {
                merge(parents, ranks, n, n - 1);
            }
            if (ranks.find(n + 1) != ranks.end()) {
                merge(parents, ranks, n, n + 1);
            }
        }
        int ans = 0;
        for (int n : nums) {
            ans = max(ans, ranks[n]);
        }
        return ans;
    }
};

File: 2024/128. Longest Consecutive Sequence.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0

        def find(parents, x):
            if parents[x] != x:
                parents[x] = find(parents, parents[x])
            return parents[x]

        def union(parents, ranks, x, y):
            x_p = find(parents, x)
            y_p = find(parents, y)
            if x_p != y_p:
                parents[y_p] = x_p
                ranks[x_p] += ranks[y_p]
        parents = defaultdict(int)
        ranks = defaultdict(int)
        for n in nums:
            parents[n] = n
            ranks[n] = 1
        for n in nums:
            if n - 1 in parents:
                union(parents, ranks, n - 1, n)
            if n + 1 in parents:
                union(parents, ranks, n + 1, n)
        return max(ranks.values())


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        count = defaultdict(int)
        ans = 0
        for n in nums:
            left = count[n - 1]
            right = count[n + 1]
            total = left + right + 1
            ans = max(ans, total)
            count[n] = total
            count[n - left] = total
            count[n + right] = total
        return ans
